js/src/logger.ts

```diff
@@ -2348,6 +2348,47 @@ export function traced<IsAsyncFlush extends boolean = false, R = void>(
   }
 }
 
+export function trace<
+  F extends (...args: any[]) => any,
+  IsAsyncFlush extends boolean = false,
+>(
+  fn: F,
+  args?: StartSpanArgs & SetCurrentArg & AsyncFlushArg<IsAsyncFlush>,
+): IsAsyncFlush extends false
+  ? (...args: Parameters<F>) => Promise<Awaited<ReturnType<F>>>
+  : F {
+  const spanArgs: typeof args = {
+    name: fn.name,
+    type: "function",
+    ...args,
+  };
+  if (args?.asyncFlush) {
+    return ((...fnArgs: Parameters<F>) =>
+      traced((span) => {
+        span.log({ input: fnArgs });
+        const output = fn(...fnArgs);
+        if (output instanceof Promise) {
+          output.then((result) => span.log({ output: result }));
+        } else {
+          span.log({ output: output });
+        }
+        return output;
```

Comment:
It feels weird to chain something to a promise (on line 2371) and then return the un-chained object. Because now someone who chains something to the output promise will be forking the promise chain rather than waiting for `span.log` to finish, which feels weird.

Maybe it should read
```
if (output instanceof Promise) {
  return (async () => {
    const result = await output;
    span.log({ output: result });
    return result;
  })();
} else {
  span.log({ output });
  return output;
}
```

Reply:
That works. I was kind of thinking you _might_ want that, but now that I think about it, promises (unlike streams) are nothing until they resolve. so we might as well combine